### ISSUE TITLE
Query nested field not working

### ISSUE THREAD

**Issue by ReneHollander** [alexklibisz/elastiknn#60]
I have the following mapping for my index:
```
{
  "properties": {
    "id": {
      "type": "long"
    },
    "features": {
      "properties": {
        "color": {
          "type": "elastiknn_dense_float_vector",
          "elastiknn": {
            "dims": 108,
            "model": "lsh",
            "similarity": "l2",
            "bands": 99,
            "rows": 1,
            "width": 3
          }
        }
      }
    }
  }
}
```
I used the Bulk API to insert my data (around 3 million features)
```
"index": {}}
{"id": 1, "features": {"color": {"values": [0.20392156862745098, 0.21176470588235294, 0.25098039215686274, 0.26666666666666666, 0.2784313725490196, 0.2235294117647059, 0.047058823529411764, 0.0392156862745098, 0.03529411764705882, 0.023529411764705882, 0.34509803921568627, 0.2235294117647059, 0.21568627450980393, 0.2235294117647059, 0.3137254901960784, 0.00392156862745098, 0.32941176470588235, 0.25098039215686274, 0.22745098039215686, 0.26666666666666666, 0.00784313725490196, 0.011764705882352941, 0.34901960784313724, 0.2627450980392157, 0.21176470588235294, 0.21176470588235294, 0.21176470588235294, 0.21176470588235294, 0.21176470588235294, 0.21568627450980393, 0.23529411764705882, 0.21568627450980393, 0.21568627450980393, 0.21568627450980393, 0.21568627450980393, 0.17647058823529413, 0.6, 0.4745098039215686, 0.25098039215686274, 0.24705882352941178, 0.19607843137254902, 0.27450980392156865, 0.25882352941176473, 0.19215686274509805, 0.13725490196078433, 0.2901960784313726, 0.1843137254901961, 0.19215686274509805, 0.3137254901960784, 0.23529411764705882, 0.1568627450980392, 0.3176470588235294, 0.19215686274509805, 0.15294117647058825, 0.34901960784313724, 0.19607843137254902, 0.33725490196078434, 0.5098039215686274, 0.24705882352941178, 0.13333333333333333, 0.5725490196078431, 0.6, 0.5215686274509804, 0.4627450980392157, 0.5725490196078431, 0.5607843137254902, 0.047058823529411764, 0.09803921568627451, 0.10196078431372549, 0.09803921568627451, 0.09019607843137255, 0.03529411764705882, 0.09803921568627451, 0.2901960784313726, 0.26666666666666666, 0.23921568627450981, 0.23921568627450981, 0.08627450980392157, 0.22745098039215686, 0.788235294117647, 0.9176470588235294, 0.8352941176470589, 0.42745098039215684, 0.10196078431372549, 0.28627450980392155, 0.7725490196078432, 0.5450980392156862, 0.44313725490196076, 0.28627450980392155, 0.10196078431372549, 0.1568627450980392, 0.2980392156862745, 0.26666666666666666, 0.24705882352941178, 0.20784313725490197, 0.058823529411764705, 0.054901960784313725, 0.19607843137254902, 0.17254901960784313, 0.1607843137254902, 0.1843137254901961, 0.06274509803921569, 0.08235294117647059, 0.12156862745098039, 0.11372549019607843, 0.12156862745098039, 0.13333333333333333, 0.10980392156862745]}}}
{"index": {}}
{"id": 2, "features": {"color": {"values": [0.0392156862745098, 0.03529411764705882, 0.027450980392156862, 0.03137254901960784, 0.0392156862745098, 0.0392156862745098, 0.0392156862745098, 0.027450980392156862, 0.0196078431372549, 0.027450980392156862, 0.0392156862745098, 0.0392156862745098, 0.0392156862745098, 0.0196078431372549, 0.023529411764705882, 0.023529411764705882, 0.03137254901960784, 0.043137254901960784, 0.03137254901960784, 0.023529411764705882, 0.027450980392156862, 0.023529411764705882, 0.027450980392156862, 0.043137254901960784, 0.03137254901960784, 0.023529411764705882, 0.023529411764705882, 0.023529411764705882, 0.03529411764705882, 0.043137254901960784, 0.027450980392156862, 0.023529411764705882, 0.023529411764705882, 0.023529411764705882, 0.03529411764705882, 0.043137254901960784, 0.4392156862745098, 0.4392156862745098, 0.5215686274509804, 0.49019607843137253, 0.36470588235294116, 0.39215686274509803, 0.34509803921568627, 0.5019607843137255, 0.48627450980392156, 0.4980392156862745, 0.3058823529411765, 0.30980392156862746, 0.3254901960784314, 0.3137254901960784, 0.34509803921568627, 0.2784313725490196, 0.26666666666666666, 0.30196078431372547, 0.38823529411764707, 0.33725490196078434, 0.21568627450980393, 0.396078431372549, 0.3058823529411765, 0.30196078431372547, 0.5176470588235295, 0.3843137254901961, 0.26666666666666666, 0.49019607843137253, 0.44313725490196076, 0.3137254901960784, 0.3254901960784314, 0.3686274509803922, 0.34901960784313724, 0.3215686274509804, 0.3411764705882353, 0.2823529411764706, 0.7529411764705882, 0.6901960784313725, 0.4196078431372549, 0.5843137254901961, 0.8274509803921568, 0.788235294117647, 0.8509803921568627, 0.5294117647058824, 0.5411764705882353, 0.5607843137254902, 0.8941176470588236, 0.8745098039215686, 0.8588235294117647, 0.7725490196078432, 0.7843137254901961, 0.8352941176470589, 0.8627450980392157, 0.8941176470588236, 0.6980392156862745, 0.7098039215686275, 0.8980392156862745, 0.6509803921568628, 0.792156862745098, 0.8980392156862745, 0.41568627450980394, 0.5725490196078431, 0.7372549019607844, 0.47058823529411764, 0.6039215686274509, 0.7607843137254902, 0.6666666666666666, 0.5529411764705883, 0.5725490196078431, 0.5882352941176471, 0.6078431372549019, 0.48627450980392156]}}}
```

I tried the following query with no success:
```
% curl -X GET http://localhost:9200/posts/_search -H "Content-type: application/json" -d @-
{
    "query": {
        "elastiknn_nearest_neighbors": {
            "field": "features.color", 
            "vec": {
                "true_indices": [0.00784313725490196, 0.30196078431372547, 0.34901960784313724, 0.3411764705882353, 0.23137254901960785, 0.0392156862745098, 0.01568627450980392, 0.00392156862745098, 0.00392156862745098, 0.011764705882352941, 0.011764705882352941, 0.027450980392156862, 0.03137254901960784, 0.2549019607843137, 0.2549019607843137, 0.34901960784313724, 0.28627450980392155, 0.3058823529411765, 0.011764705882352941, 0.3411764705882353, 0.25098039215686274, 0.24313725490196078, 0.28627450980392155, 0.011764705882352941, 0.011764705882352941, 0.0196078431372549, 0.011764705882352941, 0.011764705882352941, 0.01568627450980392, 0.0196078431372549, 0.20392156862745098, 0.33725490196078434, 0.2823529411764706, 0.00784313725490196, 0.23921568627450981, 0.24313725490196078, 0.19215686274509805, 0.058823529411764705, 0.11764705882352941, 0.1411764705882353, 0.1843137254901961, 0.03529411764705882, 0.14901960784313725, 0.10588235294117647, 0.20392156862745098, 0.37254901960784315, 0.26666666666666666, 0.10980392156862745, 0.18823529411764706, 0.10980392156862745, 0.2784313725490196, 0.27058823529411763, 0.12549019607843137, 0.0392156862745098, 0.27058823529411763, 0.21176470588235294, 0.36470588235294116, 0.30196078431372547, 0.17647058823529413, 0.24313725490196078, 0.30196078431372547, 0.27058823529411763, 0.20392156862745098, 0.17647058823529413, 0.16470588235294117, 0.18823529411764706, 0.0784313725490196, 0.0196078431372549, 0.023529411764705882, 0.13333333333333333, 0.054901960784313725, 0.03137254901960784, 0.43137254901960786, 0.5450980392156862, 0.43529411764705883, 0.5294117647058824, 0.36470588235294116, 0.803921568627451, 0.39215686274509803, 0.403921568627451, 0.2901960784313726, 0.6627450980392157, 0.3803921568627451, 0.7568627450980392, 0.3137254901960784, 0.2823529411764706, 0.16862745098039217, 0.3607843137254902, 0.2196078431372549, 0.5019607843137255, 0.3058823529411765, 0.3137254901960784, 0.12941176470588237, 0.16862745098039217, 0.1568627450980392, 0.5803921568627451, 0.25882352941176473, 0.6705882352941176, 0.6862745098039216, 0.6705882352941176, 0.7372549019607844, 0.7450980392156863, 0.6627450980392157, 0.796078431372549, 0.9058823529411765, 0.7137254901960784, 0.8627450980392157, 0.9137254901960784],
                "total_indices": 108
            },
            "model": "lsh",
            "similarity": "l2", 
            "candidates": 50
        }
    }
}
{"error":{"root_cause":[{"type":"","reason":"Attempt to decode value on failed cursor: DownField(index),DownField(vec)"}],"type":"","reason":"Attempt to decode value on failed cursor: DownField(index),DownField(vec)"},"status":500}
```

It's my first time using elasticsearch, so I have no idea how to query things...

**Comment by ReneHollander** [comment]
Looking into it further, it seems there is a error in the documentation: https://github.com/alexklibisz/elastiknn/blame/d86eecc7bb319cd8a1d6b9aa6279f299cec2e556/docs/pages/api.md#L590
This can be easily overlooked if you copy paste from the docs, instead of reading it properly...

My query now looks like this, but still finishes with an error:
```
{
    "query": {
        "elastiknn_nearest_neighbors": {
            "field": "features.color", 
            "vec": {
                "values": [0.00784313725490196, 0.30196078431372547, 0.34901960784313724, 0.3411764705882353, 0.23137254901960785, 0.0392156862745098, 0.01568627450980392, 0.00392156862745098, 0.00392156862745098, 0.011764705882352941, 0.011764705882352941, 0.027450980392156862, 0.03137254901960784, 0.2549019607843137, 0.2549019607843137, 0.34901960784313724, 0.28627450980392155, 0.3058823529411765, 0.011764705882352941, 0.3411764705882353, 0.25098039215686274, 0.24313725490196078, 0.28627450980392155, 0.011764705882352941, 0.011764705882352941, 0.0196078431372549, 0.011764705882352941, 0.011764705882352941, 0.01568627450980392, 0.0196078431372549, 0.20392156862745098, 0.33725490196078434, 0.2823529411764706, 0.00784313725490196, 0.23921568627450981, 0.24313725490196078, 0.19215686274509805, 0.058823529411764705, 0.11764705882352941, 0.1411764705882353, 0.1843137254901961, 0.03529411764705882, 0.14901960784313725, 0.10588235294117647, 0.20392156862745098, 0.37254901960784315, 0.26666666666666666, 0.10980392156862745, 0.18823529411764706, 0.10980392156862745, 0.2784313725490196, 0.27058823529411763, 0.12549019607843137, 0.0392156862745098, 0.27058823529411763, 0.21176470588235294, 0.36470588235294116, 0.30196078431372547, 0.17647058823529413, 0.24313725490196078, 0.30196078431372547, 0.27058823529411763, 0.20392156862745098, 0.17647058823529413, 0.16470588235294117, 0.18823529411764706, 0.0784313725490196, 0.0196078431372549, 0.023529411764705882, 0.13333333333333333, 0.054901960784313725, 0.03137254901960784, 0.43137254901960786, 0.5450980392156862, 0.43529411764705883, 0.5294117647058824, 0.36470588235294116, 0.803921568627451, 0.39215686274509803, 0.403921568627451, 0.2901960784313726, 0.6627450980392157, 0.3803921568627451, 0.7568627450980392, 0.3137254901960784, 0.2823529411764706, 0.16862745098039217, 0.3607843137254902, 0.2196078431372549, 0.5019607843137255, 0.3058823529411765, 0.3137254901960784, 0.12941176470588237, 0.16862745098039217, 0.1568627450980392, 0.5803921568627451, 0.25882352941176473, 0.6705882352941176, 0.6862745098039216, 0.6705882352941176, 0.7372549019607844, 0.7450980392156863, 0.6627450980392157, 0.796078431372549, 0.9058823529411765, 0.7137254901960784, 0.8627450980392157, 0.9137254901960784]
            },
            "model": "lsh",
            "similarity": "l2", 
            "candidates": 50
        }
    }
}
{"error":{"root_cause":[{"type":"query_shard_exception","reason":"failed to create query: { }","index_uuid":"yalL9pznRLKIkjjgwVBtiQ","index":"posts"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"posts","node":"VJH4aR0HRp6dWr3e8pReAA","reason":{"type":"query_shard_exception","reason":"failed to create query: { }","index_uuid":"yalL9pznRLKIkjjgwVBtiQ","index":"posts","caused_by":{"type":"unchecked_execution_exception","reason":"java.lang.RuntimeException: Failed to retrieve mapping at index [posts] field [features.color]","caused_by":{"type":"runtime_exception","reason":"Failed to retrieve mapping at index [posts] field [features.color]","caused_by":{"type":"","reason":"Attempt to decode value on failed cursor: DownField(type)"}}}}}]},"status":400}
````

**Comment by alexklibisz** [comment]
@ReneHollander Hi, thanks for finding this. I had it working at one point but I actually don't have any tests for nested fields so I probably broke it at some point. I'll plan to look into this and expand the test suite soon.

**Comment by ReneHollander** [comment]
Thanks, for now I moved it out and successfully ran queries on my dataset.

Are you also on planning to write some docs on how to select values for `rows`, `bands` and `width`? I know that the docs mention in which way query performance will go for increasing/decreasing values. Especially how they relate to the size of your dataset. I basically tried it with the values in the API Example, and inserted around 3M documents. Queries took around 5 to 10 seconds on one node (-Xmx2048M for elasticsearch). Not sure if I can exepct more on one node, or if I can achieve better with more tuning. It may make sense to open another issue for that to help others get a quick start with selecting appropriate values :)

**Comment by alexklibisz** [comment]
Those parameters depend a lot on your dataset and your desired performance
(i.e. recall vs latency trade-off). I'm actually working on some more
comprehensive benchmarking with several datasets this weekend and hope to
publish the results this coming week. I think those results will probably
be a good starting point for picking parameters.

**Comment by alexklibisz** [comment]
@ReneHollander Also quick note on performance: the general unit of parallelism in Elasticsearch is the shard. Latency should be roughly proportional to the number of shards in your index (i.e. 5 shards is ~5x faster than 1 shard, assuming you have >= 5 threads). Eventually you will want to spread shards across nodes for replication and further parallelism.

**Comment by ReneHollander** [comment]
Thanks for the reminder with the sharding. After setting it properly things got much better.

After that I put together a python script over two days I let it run over different combinations of bands, rows and width and got a good result that achieves around 100ms latency with almost perfect recall and precision.

**Comment by alexklibisz** [comment]
Great! Are you using the published python client? I haven't published much
documentation yet but its published to pypi:
https://pypi.org/project/elastiknn-client/ so you can pip install it.

**Comment by ReneHollander** [comment]
For now I have just used called the REST API directly, as I did the testing manually at first, and then built the script from there.
The application I intend to use this in is in Python, so I will likely use it when I get o that stage.

**Comment by alexklibisz** [comment]
This is fixed in #106 .
@ReneHollander If you're still using the plugin, please have a look at the latest release.
Thanks!
